### PR TITLE
Remove solver heuristics

### LINKED ISSUE #553: Failed in running cr.tl.lineages() on Ubuntu 20.04 LTS
Hello there!

I created a clean virtual environment with conda and installed cell rank by pip install cellrank in it. All the upstream processing with scvelo is superb, however I cannot proceed with the absorption probability calculation by cr.tl.lineages(adata). It didn't show any error message but just stopped responding without any progress. I used to run cell rank on macos and it was fine but since I changed my work on a linux based station recently, I could not figure out why.  

Computing lineage probabilities towards terminal states
Computing absorption probabilities

0%
0/60 [00:00<?, ?/s]

---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-13-f2982e8f2723> in <module>
      1 #absorption probabilities
----> 2 cr.tl.lineages(adata)

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/cellrank/tl/_lineages.py in lineages(adata, backward, copy, return_estimator, **kwargs)
     84 
     85     # compute the absorption probabilities
---> 86     mc.compute_absorption_probabilities(**kwargs)
     87 
     88     logg.info(f"Adding lineages to `adata.obsm[{lin_key.s!r}]`\n    Finish", time=start)

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/cellrank/tl/estimators/_base_estimator.py in compute_absorption_probabilities(self, keys, check_irreducibility, solver, use_petsc, time_to_absorption, n_jobs, backend, show_progress_bar, tol, preconditioner)
    472 
    473         # solve the linear system of equations
--> 474         mat_x = _solve_lin_system(
    475             q,
    476             s,

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/cellrank/tl/_linear_solver.py in _solve_lin_system(mat_a, mat_b, solver, use_petsc, preconditioner, n_jobs, backend, tol, use_eye, show_progress_bar)
    485         )
    486 
--> 487         mat_x, n_converged = parallelize(
    488             _solve_many_sparse_problems,
    489             mat_b,

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/cellrank/ul/_parallelize.py in wrapper(*args, **kwargs)
     98             pbar, queue, thread = None, None, None
     99 
--> 100         res = jl.Parallel(n_jobs=n_jobs, backend=backend)(
    101             jl.delayed(callback)(
    102                 *((i, cs) if use_ixs else (cs,)),

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/parallel.py in __call__(self, iterable)
   1039             # remaining jobs.
   1040             self._iterating = False
-> 1041             if self.dispatch_one_batch(iterator):
   1042                 self._iterating = self._original_iterator is not None
   1043 

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    857                 return False
    858             else:
--> 859                 self._dispatch(tasks)
    860                 return True
    861 

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/parallel.py in _dispatch(self, batch)
    775         with self._lock:
    776             job_idx = len(self._jobs)
--> 777             job = self._backend.apply_async(batch, callback=cb)
    778             # A job can complete so quickly than its callback is
    779             # called before we get here, causing self._jobs to

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    206     def apply_async(self, func, callback=None):
    207         """Schedule a func to be run"""
--> 208         result = ImmediateResult(func)
    209         if callback:
    210             callback(result)

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/_parallel_backends.py in __init__(self, batch)
    570         # Don't delay the application, to avoid keeping the input
    571         # arguments in memory
--> 572         self.results = batch()
    573 
    574     def get(self):

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/parallel.py in __call__(self)
    260         # change the default number of processes to -1
    261         with parallel_backend(self._backend, n_jobs=self._n_jobs):
--> 262             return [func(*args, **kwargs)
    263                     for func, args, kwargs in self.items]
    264 

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/joblib/parallel.py in <listcomp>(.0)
    260         # change the default number of processes to -1
    261         with parallel_backend(self._backend, n_jobs=self._n_jobs):
--> 262             return [func(*args, **kwargs)
    263                     for func, args, kwargs in self.items]
    264 

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/cellrank/tl/_linear_solver.py in _solve_many_sparse_problems(mat_b, mat_a, solver, tol, queue)
    246     for b in mat_b:
    247         # actually call the solver for the current sub-problem
--> 248         x, info = solver(mat_a, b.toarray().flatten(), tol=tol, x0=None, **kwargs)
    249 
    250         # append solution and info

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/scipy/sparse/linalg/isolve/iterative.py in gmres(A, b, x0, tol, restart, maxiter, M, callback, restrt, atol, callback_type)

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/scipy/_lib/_threadsafety.py in caller(func, *a, **kw)
     42         def caller(func, *a, **kw):
     43             with self:
---> 44                 return func(*a, **kw)
     45         return scipy._lib.decorator.decorate(func, caller)
     46 

~/miniconda3/envs/cr1.3/lib/python3.8/site-packages/scipy/sparse/linalg/isolve/iterative.py in gmres(A, b, x0, tol, restart, maxiter, M, callback, restrt, atol, callback_type)
    598         olditer = iter_
    599         x, iter_, presid, info, ndx1, ndx2, sclr1, sclr2, ijob = \
--> 600            revcom(b, x, restrt, work, work2, iter_, presid, info, ndx1, ndx2, ijob, ptol)
    601         if callback_type == 'x' and iter_ != olditer:
    602             callback(x)

KeyboardInterrupt: 

#### Versions:
Package             Version
------------------- ---------
anndata             0.7.5
argon2-cffi         20.1.0
async-generator     1.10
attrs               20.3.0
backcall            0.2.0
bleach              3.3.0
cellrank            1.3.0
certifi             2020.12.5
cffi                1.14.5
click               7.1.2
cycler              0.10.0
decorator           4.4.2
defusedxml          0.7.1
docrep              0.3.2
entrypoints         0.3
environment-kernels 1.1.1
future              0.18.2
future-fstrings     1.2.0
get-version         2.1
h5py                3.2.1
ipykernel           5.5.3
ipython             7.22.0
ipython-genutils    0.2.0
ipywidgets          7.6.3
jedi                0.18.0
Jinja2              2.11.3
joblib              1.0.1
jsonschema          3.2.0
jupyter-client      6.1.12
jupyter-core        4.7.1
jupyterlab-pygments 0.1.2
jupyterlab-widgets  1.0.0
kiwisolver          1.3.1
legacy-api-wrap     1.2
llvmlite            0.35.0
loompy              3.0.6
MarkupSafe          1.1.1
matplotlib          3.4.1
mistune             0.8.4
natsort             7.1.1
nbclient            0.5.3
nbconvert           6.0.7
nbformat            5.1.2
nest-asyncio        1.5.1
networkx            2.5
notebook            6.3.0
numba               0.52.0
numexpr             2.7.3
numpy               1.20.2
numpy-groupies      0.9.13
packaging           20.9
pandas              1.2.3
pandocfilters       1.4.3
parso               0.8.2
patsy               0.5.1
pexpect             4.8.0
pickleshare         0.7.5
Pillow              8.1.2
pip                 21.0.1
progressbar2        3.53.1
prometheus-client   0.9.0
prompt-toolkit      3.0.18
ptyprocess          0.7.0
pycparser           2.20
pygam               0.8.0
Pygments            2.8.1
pygpcca             1.0.2
pyparsing           2.4.7
pyrsistent          0.17.3
python-dateutil     2.8.1
python-igraph       0.9.1
python-utils        2.5.6
pytz                2021.1
pyzmq               22.0.3
scanpy              1.7.1
scikit-learn        0.24.1
scipy               1.6.2
scvelo              0.2.3
seaborn             0.11.1
Send2Trash          1.5.0
setuptools          54.2.0
sinfo               0.3.1
six                 1.15.0
statsmodels         0.12.2
stdlib-list         0.8.0
tables              3.6.1
terminado           0.9.4
testpath            0.4.4
texttable           1.6.3
threadpoolctl       2.1.0
tornado             6.1
tqdm                4.59.0
traitlets           5.0.5
typing-extensions   3.7.4.3
tzlocal             2.1
umap-learn          0.4.6
wcwidth             0.2.5
webencodings        0.5.1
wheel               0.30.0
widgetsnbextension  3.5.1
wrapt               1.12.1

### PR DESCRIPTION
**Pull request by michalk8**
**IMPORTANT: Please search among the [Pull request](../pulls) before creating one.**

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)

## Description

Remove heuristics which determine whether to use PETSc and direct solver.
Use PETSc when installed by default and `gmres` regardless of the problem size.

## How has this been tested?

Current tests should be sufficient.

## Closes

closes #553

### CODE REVIEW

**Comment by Marius1311** [comment]
Can you just quickly give me the rationale for this please? Is this to make the method more predictable?

**Comment by michalk8** [comment]
> Can you just quickly give me the rationale for this please? Is this to make the method more predictable?

Rationale:
- PETSc >= LAPACK/ARPACK in many cases (c.f. eigs problem we had in pygpcca)
- direct solver (in scipy) seems to have cause the issue in #553 , we could go back to it, but I'd improve the heuristics (simpliy) + keep the PETSc change

**Comment by Marius1311** [comment]
Ok, nice, feel free to merge when tests pass

File: cellrank/tl/estimators/_base_estimator.py [hunk 1]
```diff
@@ -341,12 +341,9 @@ def compute_absorption_probabilities(
 
             Information on the :mod:`scipy` iterative solvers can be found in :func:`scipy.sparse.linalg` or for
             :mod:`petsc4py` solver `here <https://www.mcs.anl.gov/petsc/documentation/linearsolvertable.html>`__.
-
-            If is `None`, the solver is chosen automatically, depending on the problem size.
         use_petsc
             Whether to use solvers from :mod:`petsc4py` or :mod:`scipy`. Recommended for large problems.
-            If `None`, it is determined automatically. If no installation is found, defaults to
-            :func:`scipy.sparse.linalg.gmres`.
+            If no installation is found, defaults to :func:`scipy.sparse.linalg.gmres`.
```
Comment (Marius1311) [inline cellrank/tl/estimators/_base_estimator.py:346]: Did you check whether that makes it a lot slower on small problems? I know for sure that the direction method **will be* faster for small problems, the question is *how much*.

**Review by Marius1311 — COMMENTED** [review]
Just not sure why we need to use gmres for small problems.

**Comment by Marius1311** [comment]
We can discuss this tomorrow also. It's minor.

File: cellrank/tl/estimators/_base_estimator.py (hunk 1 above)
Comment (michalk8) [reply cellrank/tl/estimators/_base_estimator.py:346]:
```
# keys are subsets of the pancreas data, values are (mean,std) for 10 random subsets
PETSc + gmres
{100: (0.011416912078857422, 0.004280086204327923),
 250: (0.018981313705444335, 0.003186679585923787),
 500: (0.021230626106262206, 0.0012997817509389086),
 1000: (0.057153940200805664, 0.005951826256232277),
 1500: (0.09224464893341064, 0.011383153019094908),
 2000: (0.1188497543334961, 0.0013607115998565049),
 2500: (0.1569295644760132, 0.0022418019388140424)}

PETSc + direct
{100: (0.005582404136657715, 0.0013073040658704105),
 250: (0.007641458511352539, 0.0016195266670197747),
 500: (0.012239551544189453, 0.0015909780405909395),
 1000: (0.028612852096557617, 0.002168457370086139),
 1500: (0.07074854373931885, 0.012893574633071328),
 2000: (0.1430955410003662, 0.01825136893836883),
 2500: (0.2457975149154663, 0.024587187138362544)}

scipy + direct
{100: (0.0058349609375, 0.0017682622847889555),
 250: (0.014855360984802246, 0.01063481691736127),
 500: (0.01383976936340332, 0.0019364345745918484),
 1000: (0.033646798133850096, 0.009655263707504665),
 1500: (0.16855034828186036, 0.11022524525949293),
 2000: (0.1577007532119751, 0.08393314372050473),
 2500: (0.2498032808303833, 0.04451559792479094)}
```

Comment (Marius1311) [reply cellrank/tl/estimators/_base_estimator.py:346]: really well done @michalk8, thank you very much for looking into this and for running these comparisons!

**Pull request merged**